library/std/src/os/unix/io/mod.rs

```diff
@@ -26,20 +26,30 @@
 //! that they don't outlive the resource they point to. These are safe to
 //! use. `BorrowedFd` values may be used in APIs which provide safe access to
 //! any system call except for:
+//!
 //!  - `close`, because that would end the dynamic lifetime of the resource
 //!    without ending the lifetime of the file descriptor.
+//!
 //!  - `dup2`/`dup3`, in the second argument, because this argument is
 //!    closed and assigned a new resource, which may break the assumptions
 //!    other code using that file descriptor.
-//! This list doesn't include `mmap`, since `mmap` does do a proper borrow of
-//! its file descriptor argument. That said, `mmap` is unsafe for other
-//! reasons: it operates on raw pointers, and it can have undefined behavior if
-//! the underlying storage is mutated. Mutations may come from other processes,
-//! or from the same process if the API provides `BorrowedFd` access, since as
-//! mentioned earlier, `BorrowedFd` values may be used in APIs which provide
-//! safe access to any system call. Consequently, code using `mmap` and
-//! presenting a safe API must take full responsibility for ensuring that safe
-//! Rust code cannot evoke undefined behavior through it.
+//!
+//! `BorrowedFd` values may be used in APIs which provide safe access to `dup`
+//! system calls, so types implementing `AsFd` or `From<OwnedFd>` should not
+//! assume they always have exclusive access to the underlying file
+//! description.
```

Comment:
I'm wondering how useful this statement is.
Any code that does read raw bytes and then transmutes them into some `repr(C)` type may end up reading the wrong data when another thread concurrently messes with the seek position on the file descriptor. Even checking some version header, endianness and so on won't help against that. One would have to hash the payload bytes themselves before transmuting which I assume many state serialization schemes don't do for performance reasons.

This is similar to the mmap problem but more subtle because merely seeking seems innocuous enough on its own.

I know that due to `File::try_clone` this already is a problem with any code that takes a `File`. But since we're trying to define IO-safety here it kind of seems like a subtle hole that few people will defend against in practice and I'm not sure the documentation is enough.

Reply:
If we say it's valid to have `as_filelike_view::<File>()`, then users can do `try_clone()` from any `BorrowedFd`. If we don't want that, then we need to figure out somewhere where we can draw a line.

Data read from files should almost never be trusted in a Rust safety sense, so arguably this is a correctness hazard rather than a safety hazard. It'd still be nice to do something about it, but it's not clear what we could do.

Reply:
It doesn't even take a view. You can take a `File`, clone it, then you have two `File`s and turn each into an `OwnedFd` where each then can be passed to some `From<OwnedFd>` type.

> Data read from files should almost never be trusted in a Rust safety sense

Most of the examples I can think of are only adversarial ones. A library that is fairly defensive or at the very minimum does some sanity checks but then after doing the sanity checks keeps using the file in more performance-oriented ways that would only be subverted if a) some external process modifies the file (this is no worse than the `/proc/self/mem` situation) or b) some other part of the rust program plays with the shared file description (this is not (much?) worse than `File::try_clone`).

So those safety/complexity/performance tradeoffs involved aren't new. But not we're writing explicit documentation about them and I think the proposed documentation doesn't quite cover those edge-cases.

Imo it is quite reasonable for many programs (e.g. ones that aren't suid executables, sandboxes or taking data from the internet) that care about performance to have something this somewhere:

```rust
fn rehydrate_state(&mut self, input: impl AsFd) -> io::Result<()> {
   let file = input.as_file_like_view<File>();
   // do resasonable checks here. magic number, checksummed trailer, file length, endianness, struct version
   // maybe even check that the fd and inode is read-only
   self.validate_data(&file)?;
   // PogramState is a repr(C) struct explicitly designed to be dumpable. parts get loaded via MaybeUninit
   let program_state: MaybeUninit<ProgramState> = unsafe { self.deserialize(&file)?; 
   // we did pre-validation, we didn't unexpectedly hit EOF or read errors
   self.state = unsafe { program_state.assume_init() };
}
```

The `AsFd` is used because the snapshot might be passed in as stdin or from a fork as a numbered file descriptor.
If we want to use `mmap` we might require `OwnedFd` instead of `impl AsFd`, but still, we accept an arbitrary file descriptor here.

And then it is technically unsafe because inside `deserialize` it's probably doing multiple reads that could get confused by concurrent seeks. But it looks like a sane use to me under non-adversarial conditions where things don't randomly write into your filesystem or fiddle with your file descriptions.

Reply:
Are you advocating for changing the design to make `dup` unsafe? If so, could I ask you to make a specific proposal? If not, would you be able to propose some wording, even if it's just a starting point, for what the documentation might say here?

Reply:
I realized that my example is wrong, an `impl AsFd` already would be too permissive because it would allow multiple threads to read from the same file which of course would already affect the seek position.
We'd at least need some *pretense* of exclusive access to make it even remotely sane. I'll update the example.

I think we should ask for opinions on zulip whether internally-unsafe code that takes an `OwnedFd` must always be an `unsafe fn` because it might still be a shared file description.

> Are you advocating for changing the design to make dup unsafe?

I'd love to, but there are more subtle and useful ways to dup file descriptors such as sending them over unix sockets or sharing them across forks. And `File::try_clone` already is stable anyway.

Reply:
Between separate opens of the same file, existing safe ways to duplicate a file, and the fact that file descriptors could refer to many different things, I don't think we can do anything more than we're doing here.

Reply:
In particular, code reading from a file should be reading data then checking it (to avoid TOCTTOU).